xline/src/server/watch_server.rs

```diff
@@ -406,12 +433,95 @@ mod test {
         };
         req_tx1.send(Ok(w_req.clone())).await?;
         req_tx2.send(Ok(w_req.clone())).await?;
-        tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+        sleep(Duration::from_secs(1)).await;
         for count in collection.lock().values() {
             assert_eq!(*count, 1);
         }
         handle1.abort();
         handle2.abort();
         Ok(())
     }
+
+    #[tokio::test]
+    async fn test_watch_prev_kv() {
+        let index = Arc::new(Index::new());
+        let db = DB::open(&StorageConfig::Memory).unwrap();
+        let header_gen = Arc::new(HeaderGenerator::new(0, 0));
+        let lease_collection = Arc::new(LeaseCollection::new(0));
+        let next_id_gen = Arc::new(WatchIdGenerator::new(1));
+        let (kv_update_tx, kv_update_rx) = mpsc::channel(CHANNEL_SIZE);
+        let kv_store = Arc::new(KvStore::new(
+            kv_update_tx,
+            lease_collection,
+            Arc::clone(&header_gen),
+            Arc::clone(&db),
+            index,
+        ));
+        let shutdown_trigger = Arc::new(event_listener::Event::new());
+        let kv_watcher = KvWatcher::new_arc(
+            Arc::clone(&kv_store),
+            kv_update_rx,
+            shutdown_trigger,
+            Duration::from_millis(10),
+        );
+        put(&kv_store, &db, "foo", "old_bar", 2).await;
+        put(&kv_store, &db, "foo", "bar", 3).await;
+
+        let (req_tx, req_rx) = mpsc::channel(CHANNEL_SIZE);
+        let req_stream = ReceiverStream::new(req_rx);
+        let create_watch_req = move |watch_id: WatchId, prev_kv: bool| WatchRequest {
+            request_union: Some(RequestUnion::CreateRequest(WatchCreateRequest {
+                watch_id,
+                key: "foo".into(),
+                start_revision: 3,
+                prev_kv,
+                ..Default::default()
+            })),
+        };
+        req_tx.send(Ok(create_watch_req(1, false))).await.unwrap();
+        req_tx.send(Ok(create_watch_req(2, true))).await.unwrap();
+        let (res_tx, mut res_rx) = mpsc::channel(CHANNEL_SIZE);
+        let _hd = tokio::spawn(WatchServer::<DB>::task(
+            Arc::clone(&next_id_gen),
+            Arc::clone(&kv_watcher),
+            res_tx,
+            req_stream,
+            Arc::clone(&header_gen),
+        ));
+
+        for _ in 0..4 {
+            let watch_res = res_rx.recv().await.unwrap().unwrap();
+            if watch_res.events.is_empty() {
+                // create response
```

Comment:
create response?

Reply:
Fixed, it should be `WatchCreateResponse`